src/include/duckdb/main/client_context_state.hpp

```diff
@@ -78,4 +80,63 @@ class ClientContextState {
 	}
 };
 
+class RegisteredStateIterator {
+public:
+	using iterator_type = unordered_map<string, shared_ptr<ClientContextState>>::iterator;
+	using value_type = unordered_map<string, shared_ptr<ClientContextState>>::value_type;
+
+public:
+	RegisteredStateIterator(RegisteredStateManager &manager, iterator_type iterator_p, idx_t state_version);
+
+	reference<RegisteredStateManager> manager_ref;
+	iterator_type iterator;
+	idx_t state_version;
+
+public:
+	RegisteredStateIterator &operator++();
+	bool operator!=(const RegisteredStateIterator &other) const;
+	value_type &operator*() const;
+};
+
+class RegisteredStateManager {
```

Comment:
also provide a Remove method to maintain feature parity with todays registered_state?

Reply:
Done